Check timestamp/snapshot contains snapshot/targets version (#226)
theupdateframework/specification: tuf-spec.md

3. **Check for a rollback attack.**

  1. The version number of the trusted timestamp metadata file, if
    any, MUST be less than the version number of the new timestamp
    metadata file. If the new timestamp metadata version is less than the trusted
    timestamp metadata version, discard it, abort the update cycle, and
    report the potential rollback attack. In case they are equal, discard the new
    timestamp metadata and abort the update cycle. This is normal and it
    shouldn't raise any error. The reason for aborting the update process is that
    there shouldn't be any changes in the content of this, or any other metadata

>     files too, considering it has the same version as the already trusted one.

Reviewer: I believe the only change here is line wrap / formatting.  Can you omit this change?

Author: done

>   2. The new timestamp metadata file MUST only contain the description of the
>      snapshot metadata file.  If not, discard the new snapshot metadata file,
>      abort the cycle, and report the failure.

Reviewer 2: ```suggestion
  2. The new timestamp metadata file MUST only contain the description of the
     snapshot metadata file.  If not, discard the new timestamp metadata file,
     abort the cycle, and report the failure.
```

Just to be clear, here we're checking that timestamp's `meta` lists _only_ snapshot?

>   3. The version number of the snapshot metadata file in the trusted timestamp

Reviewer: Same here.  Please don't redo the line breaks as part of another change.

Author: done

    shouldn't raise any error. The reason for aborting the update process is
    that there shouldn't be any changes in the content of this, or any other
    metadata files too, considering it has the same version as the already
    trusted one.

  2. The new timestamp metadata file's [=METAFILES=] object MUST only
    contain the snapshot metadata file.  If not, discard the new timestamp
    metadata file, abort the cycle, and report the failure.

  3. The [=metapath/VERSION=] number of the snapshot metadata file in the
    trusted timestamp metadata file, if any, MUST be less than or equal to its
    [=metapath/VERSION=] number in the new timestamp metadata file.  If not,
    discard the new timestamp metadata file, abort the update cycle, and report
    the failure.

  4. If the new timestamp metadata file's [=metapath/VERSION=] number of the
    snapshot metadata file is equal to the [=metapath/VERSION=] numbers in the
    trusted snapshot metadata file:

    1. The [=metapath/LENGTH=] in the new timestamp metadata file, if any, MUST
      be equal to the [=metapath/LENGTH=] in the trusted timestamp file, if any.

>       If not, discard the new timestamp metadata file, abort the cycle, and

Reviewer: Why do we discard the new timestamp metadata?  Could an attacker later roll back the timestamp file?

Author: I'm not sure if it's necessarily resisting an attack, but more about enabling the optimization where we can skip downloading metadata if we already have it locally without putting our local trusted metadata in an inconsistent state (such as in #114 and #227).

It might be helpful if we work with an example. Say we have locally trusted metadata, with Timestamp T1 and Snapshot S1, which is at version N. The remote repository is at Timestamp T1 and Snapshot S2, which also is at version N.

When we perform a client update, we will:

* Download T2
* Verify it was signed correctly
* Verify that T2's snapshot description's version >= the trusted snapshot S1's version

At this point we could decide that we we already have the trusted Snapshot S1, so we could abort the update workflow. However this would be incorrect, because it's possible that Snapshot S2 could have different content than Snapshot S1. If Timestamp T2 contains the optional length and/or hashes of Snapshot S2. This means that if we create a brand new client based off the local trusted metadata, we cannot use the length/hashes from Timestamp T2 to verify Snapshot S1. This would then force us to throw away the Snapshot S1 and download the new S2 version.

This patch tries to avoid this issue by enforcing that we cannot get into this state by treating Timestamp T2 as invalid, since it doesn't match what we already trust.

This would be especially handy if we implement transactional downloading of metadata in #69 to make sure the local metadata doesn't get into an incomplete state.

Reviewer 3: > The remote repository is at Timestamp T1 and Snapshot S2, which also is at version N.

Did you mean **T2**? You do say the client downloads **T2** below. I assume this is a typo here?

Just to make sure, do I understand correctly that in your example:
- T1 describes S1
- T2 describes S2
- T2 has a higher version than T1
- S2 has the same version as S1 
- S2 and S1 have different contents

This implies that the repo has *incorrectly* updated snapshot and timestamp, without bumping snapshot's version number, right?

Following you proposal, the client that already trusts T1 and S1 can only update, if the repo issues a new T2 (or higher) describing a new S2, with a version number higher than S1. Is that correct? If so, I think it's fair.

      report the failure.

    2. For each entry in the new timestamp metadata file's [=metapath/HASHES=]
      dictionary, if the key is present in the trusted timestamp metadata file,

>       the values MUST be equal.  If not, discard the new timestamp metadata

Reviewer: Same comment as above here.

Author: This was covered in the above comment, so lets track it there.

  in the trusted timestamp metadata.  If the versions do not match, discard the
  new snapshot metadata, abort the update cycle, and report the failure.

5. **Check for a rollback attack**.

>   1. The new snapshot metadata file MUST contain the description of the targets

Reviewer: ```suggestion
  1. The new snapshot metadata file MUST contain the version numbers of the targets
```
The listed item is the version numbers, not a description.

Reviewer 2: I assume that "description" was used as a short-hand for the version plus optional length and hashes (fields in a `METAFILES` object)? As this terminology is not used elsewhere we should certainly be more exploit here.

Reviewer 2: Perhaps we could just say
>The new snapshot metadata file's `METAFILES` object MUST list all targets metadata (top-level targets.EXT all all delegated roles) including their version numbers, optionally their length and optionally their hashes. If not, discard the new snapshot metadata file, abort the 

etc.

>      metadata file.  If not, discard the new snapshot metadata file, abort the

Reviewer: ```suggestion
     metadata files (the top-level targets.EXT and all delegated roles.)  If not, discard the 
     new snapshot metadata file, abort the
```
Clarifying that this includes all targets files.

  in the trusted timestamp metadata.  If the versions do not match, discard the
  new snapshot metadata, abort the update cycle, and report the failure.

5. **Check for a rollback attack**.

  1. The new snapshot metadata file's [=METAFILES=] object MUST contain a
    [=snapshot/METAPATH=] entry for the targets metadata file, and all delegated targets
    metadata files, if any, in the trusted snapshot metadata file.  If not,

>     discard the new snapshot metadata file, abort the cycle, and report the

Reviewer: Do we need to discard the snapshot file in this case?  Why?

Author: This was covered in the above comment, so lets track it there.
